### PR TITLE
Stdsim fixes

### LINKED ISSUE #654: Investigate potential StdSim and/or ByteBuf improvements
The `StdSim` and `ByteBuf` classes were initially created as a convenient hack for use in unit and transcript testing with little to no thought given to factors like performance.  However, ``StdSim`` is now used by *pyscript_bridge.py* and hence by the **py**, **pyscript**, and **ipy** commands.

Issues that should be investigated include the following:
1. `ByteBuf` uses an immutable `bytes` object as a buffer to store text in the inner buffer.  If a lot of data gets sent here, constantly recreating new versions might be inefficient.  We should consider replacing it with some mutable type like a `bytearray`.
1. `StdSim` should probably behave like it's inner_stream. 
    * If the `inner_stream` is a `_io.TextIOWrapper` like `sys.stdout` typically is, then `StdSim` should provide the `buffer` interface.
    * If the `inner_stream` is a _`io.BufferedWriter` like a file opened in binary mode, then no `buffer` interface would be provided.
    *  In both cases, the `write` method should behave like the inner stream.
        * In the text case, `write(`) should accept strings, and in the binary case, bytes.

These changes will support wrapping `stdout` no matter what type of stream it is.

### PR DESCRIPTION
**Pull request by kmvanbrunt**
Closes #654 
Removed support for wrapping binary streams in `ProcReader`.

These changes also improve the flushing behavior of when `pyscript` echos the output of a shell command. Since we obtain that output over binary pipes, we echo it using `StdSim.inner_stream.buffer.write()` which does not follow line buffering behavior by default. The new code will call `flush()` if the bytes being written contain a newline or carriage return and the inner stream has line buffering turned on.

This makes echoed shell command output print at the same frequency it would when run at the terminal. Commands like `ping` are a good test case for this. ex: `app('ping localhost')`

### CODE REVIEW

File: cmd2/utils.py [hunk 1]
```diff
@@ -350,6 +350,17 @@ def isatty(self) -> bool:
         else:
             return False
 
+    @property
+    def line_buffering(self) -> bool:
+        """
```
Comment (tleonhardt) [inline cmd2/utils.py]: Consider adding a unit test or two to make sure pith positive and negative cases of this method are cover d.

File: cmd2/utils.py [hunk 2]
```diff
@@ -374,14 +388,22 @@ def write(self, b: bytes) -> None:
         if self.std_sim_instance.echo:
             self.std_sim_instance.inner_stream.buffer.write(b)
 
+            # Since StdSim wraps TextIO streams, we will flush the stream if line buffering is on
+            # and the bytes being written contain a new line character. This is helpful when StdSim
+            # is being used to capture output of a shell command because it causes the output to print
+            # to the screen more often than if we waited for the stream to flush its buffer.
+            if self.std_sim_instance.line_buffering:
+                if any(newline in b for newline in ByteBuf.NEWLINES):
+                    self.std_sim_instance.flush()
```
Comment (tleonhardt) [inline cmd2/utils.py]: This flushing is a nice improvement

File: tests/test_utils.py [hunk 3]
```diff
@@ -218,6 +207,26 @@ def test_stdsim_pause_storage(stdout_sim):
     stdout_sim.buffer.write(b_str)
     assert stdout_sim.getbytes() == b''
 
+def test_stdsim_line_buffering(base_app):
```
Comment (tleonhardt) [inline tests/test_utils.py]: Thanks for adding the unit test
